[JENKINS-20052] Refactor ListView getItems

res0nance opened the pull request:
Signed-off-by: Raihaan Shouhell <raihaan.shouhell@autodesk.com>

See [JENKINS-20052](https://issues.jenkins-ci.org/browse/JENKINS-20052).

I discovered the above jira issue while poking around. This supersedes #4462.
This is in hopes to improve performance with https://github.com/jan-molak/jenkins-build-monitor-plugin.

Current implementation is ugly but I'd like feedback on if this is the right approach

### Proposed changelog entries

* Entry 1: Reduce number of calls to getItems() and iterations through the jobList in ListView
* ...

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Proposed upgrade guidelines

- List View's job status filter has been moved to be a Job View Filter. The order of this filter matters as its interaction with various other filters will affect what jobs are shown in the view. If you set the status filter to be the first filter(filtering only enabled jobs) and a later filter adds a disabled job. you will see the disabled job. If the status filter is last in the same scenario, you will not see the disabled job.

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@jglick 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

Commit:
Refactor ListView getItems

Signed-off-by: Raihaan Shouhell <raihaan.shouhell@autodesk.com>

res0nance commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -211,28 +212,57 @@ public boolean hasJobFilterExtensions() {
+                    if (itemName.equals(name)) {
+                        if (statusFilter == null || !(item instanceof ParameterizedJobMixIn.ParameterizedJob) // TODO or better to call the more generic Job.isBuildable?
+                                || ((ParameterizedJobMixIn.ParameterizedJob) item).isDisabled() ^ statusFilter)
+                            items.add(item);
```
Probably should be refactored

res0nance commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -211,28 +212,57 @@ public boolean hasJobFilterExtensions() {
         // check the filters
         Iterable<ViewJobFilter> jobFilters = getJobFilters();
-        List<TopLevelItem> allItems = new ArrayList<>(parentItems);
-        if (recurse) allItems = expand(allItems, new ArrayList<>());
```
Assumed that expand is the same as `parent.getAllItems()` in this case

Commit:
Minor Refactor

Commit:
Optimization for specific case

timja reviewed (commented):
any chance of a benchmark, before / after the change?

Also is the test coverage sufficient around here that the tests passing is enough for this change to be okay?

res0nance commented:
> any chance of a benchmark, before / after the change?
I haven't written one so no

> Also is the test coverage sufficient around here that the tests passing is enough for this change to be okay?

From what i read of the tests it seems to be sufficient. Except there doesn't seem to be any tests with `ViewJobFilter`

jglick reviewed (commented):
I am afraid I have no idea what this patch does, and have forgotten most of the context from JIRA.

res0nance commented:
@jglick This is to reduce the amount of calls to `getItems()` and refactor the function to reduce number of passes. 

The jira ticket is from 2013 so 7 years ago, hard to remember the context after so long.

I'm basically making it do this
From the JIRA:
Basically the issue is that instead of doing the obvious thing—looking up explicitly-specified names using the method that exists for that purpose, then scanning recursive items whose relative name match the regexp and adding those in—the view is scanning recursive items, adding them to the name list, then going back and scanning recursive items again and collecting those whose (relative) names match the list. And the job filters are reported to be called with the wrong arguments: the direct children of the folder.

daniel-beck commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -242,6 +264,13 @@ public boolean hasJobFilterExtensions() {
     }
 
+    // Used by getItems
+    private static void checkAddItem(Boolean statusFilter, List<TopLevelItem> items, TopLevelItem item) {
```
```suggestion
    private static void addItemMatchingStatus(Boolean statusFilter, List<TopLevelItem> items, TopLevelItem item) {
```

Plus corresponding changes.

daniel-beck commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -211,28 +212,49 @@ public boolean hasJobFilterExtensions() {
+                    //Ignore
+                }
+            }
+            //Only call getItems if there is a pattern or we have Job Filters
```
I think it would help readability if this comment had 2 specific lines of matching code, rather than it explaining the if/else if structure with a repeated statement in block blocks.

daniel-beck reviewed (commented):
Seems OK from reading it, but wow this was convoluted code before, and the many options don't make it all that much easier.

I still think the cleanup I mentioned in Jira should be done to simplify this; no reason for status check to not be a filter.

Haven't tested it, and the only case I could think of that had the potential to break things seems to be taken care of by filtering for `TopLevelItem` (Matrix/Maven projects). I _really_ hope this has some good test coverage…

I recommend this be manually tested with View Job Filters Plugin, but note there are some weird filters (the "not in another view" filter is pretty broken IIRC).

Thanks for cleaning this mess up!

Commit:
Migrate statusfilter to viewjobfilter

Commit:
Use unboxed boolean so we don't have to deal with null

daniel-beck commented:
@res0nance Wouldn't it make more sense to deliver these separately? Or do you save so much with combining them?

daniel-beck commented:
Nice. This seems quite a bit more straightforward.

I would expect the migration of the status filter to require a corresponding UI update.

And for safety, it should be mentioned in upgrade documentation.

res0nance commented:
I've noticed that this statusfilter migration will change behaviour.

Consider If the status filter happens after a jobviewfilter it can filter out certain jobs that are disabled.

daniel-beck commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -145,6 +147,9 @@ private Object readResolve() {
         initColumns();
         initJobFilters();
+        if (statusFilter != null) {
+            jobFilters.add(new StatusFilter(statusFilter));
```
This should add the `StatusFilter` first into the list to minimize behavior changes through interactions between filters.

res0nance replied:
Rewritten to be added to the first index

Commit:
UI Changes

Commit:
Merge branch 'master' of https://github.com/res0nance/jenkins into JENKINS-20052

Commit:
Add tests and use pre-auth filters

Commit:
Add test with a simple filter

res0nance commented:
I think this is done:
- I've manually tested this with a couple of filters in view-job-filters plugin.
- I've split the status filter out
- I've added more tests for both ListView and the new status filter

timja reviewed (approved):
Code looks much clearer to me and is much better tested, thanks!

res0nance commented:
Need to fix an old test as it doesn't mock the predicate forms
https://github.com/jenkinsci/jenkins/blob/69a524ab0b248593fa0ebd42dc5447db91d8d55c/core/src/test/java/hudson/model/ListViewTest.java#L52

Commit:
Remove old tests as the new tests should cover this

Commit:
Change annotation

res0nance commented:
Had to be updated due to #4604

res0nance commented:
@oleg-nenashev and @daniel-beck could you take a look at this?

timja commented:
@jenkinsci/core-pr-reviewers / @daniel-beck would you mind taking a look please?

Commit:
Merge branch 'master' into JENKINS-20052

Commit:
Only iterate if there are names or a regex pattern

res0nance commented:
Is there any way to move this forward?

MRamonLeon commented:
@res0nance Due to the size of the PR I couldn't review it today. I can reserve a dedicated time for it next Thursday. I will try to do it before, but I cannot guarantee.

MRamonLeon reviewed (approved):
I've tested manually and it performs correctly and returns the same jobs as:

`docker run ... jenkins/jenkins:2.237-alpine`

I've used a script to populate Jenkins with folders and jobs although I didn't see much difference with 1024 jobs. I leave it here for anyone interested in test the performance with better knowledge about the changes.

<details>
<summary>Script to populate folders and jobs</summary>


```
import jenkins.model.Jenkins;
import hudson.model.FreeStyleProject;
import hudson.tasks.Shell;
import com.cloudbees.hudson.plugins.folder.Folder;
import hudson.model.Project;
Jenkins jenkins = Jenkins.instance

@groovy.transform.Field
int MAX = 4; // 4^5 jobs = 1024 jobs. Be careful, bigger numbers could eat your machine

for (int i = 1; i <= MAX; i++) {
  createSubFolderPopulated(jenkins, "folder-${i}", 1);
}

Folder createSubFolderPopulated(ModifiableTopLevelItemGroup parent, String name, Integer level) throws IOException {
  Folder f = createFolder(parent, name);

  if (level == MAX) {
	for (int k = 1; k <= MAX; k++) {
      Project job = createProject(f, "${name}-${k}");
      
      //job.buildersList.add(new Shell("echo hello world from ${i}-${j}-${k}"))
      //job.save()
      //build = job.scheduleBuild2(10, new hudson.model.Cause.UserIdCause())
      //build.get() # Block until the build finishes
    }
  } else {
  
    level++;
    
    for (int i = 1; i <= MAX; i++) {
      createSubFolderPopulated(f, "${name}-${i}", level);
    }
  }
  
  return f;
}
                                  
Folder createFolder(ModifiableTopLevelItemGroup parent, String name) throws IOException {
  return (Folder) parent.createProject((TopLevelItemDescriptor) new Folder.DescriptorImpl(), name);
}


FreeStyleProject createProject(ModifiableTopLevelItemGroup parent, String name) throws IOException {
    return (FreeStyleProject) parent.createProject(FreeStyleProject.DESCRIPTOR, name);
}



long getNumberJobs() {
    Jenkins.instance.getAllItems(FreeStyleProject.class).size()
}
```
</details>

res0nance commented:
Marking this ready-for-merge

> I've used a script to populate Jenkins with folders and jobs although I didn't see much difference with 1024 jobs. I leave it here for anyone interested in test the performance with better knowledge about the changes.

Majority of the perf benefit is going to come from users who require login as this reduces the amount of Permissions checks

daniel-beck commented:
[JENKINS-62661](https://issues.jenkins-ci.org/browse/JENKINS-62661) reports data loss related to views when updating from 2.238 to 2.240. All affected views have `<statusFilter>true</statusFilter>`, so this might be related?

Also, some views are from https://plugins.jenkins.io/build-monitor-plugin/, but not all. This could indicate a backward incompatible change?

res0nance commented:
Not sure how this is incompatible. But it does sound like something is incompatible

daniel-beck commented on core/src/main/java/hudson/model/ListView.java:
```diff
@@ -145,6 +149,9 @@ private Object readResolve() {
         initColumns();
         initJobFilters();
+        if (statusFilter != null) {
+            jobFilters.add(0, new StatusFilter(statusFilter));
```
@TobiX identified this as the problem in [JENKINS-62661]( https://issues.jenkins-ci.org/browse/JENKINS-62661): `DescribableList` does not implement `#add(int,T)`.

pjwiseman commented:
Noting that changes results in Views with the StatusFilter set being removed when upgrading from <2.239 to >2.239.

Related issues: [JENKINS-62661](https://issues.jenkins-ci.org/browse/JENKINS-62661), [JENKINS-63786](https://issues.jenkins-ci.org/browse/JENKINS-63678).

MarkEWaite commented:
@res0nance see the description of the incompatibility in [JENKINS-62661](https://issues.jenkins-ci.org/browse/JENKINS-62661).  I assume we need a readResolve() implementation that handles the prior statusFilter element and converts it into the new form.
